python/ray/tests/test_tf_differentiable.py

```diff
@@ -0,0 +1,623 @@
+from functools import reduce
+
+import pytest
+import tensorflow as tf
+import numpy as np
+
+import ray
+from ray.experimental.tf_utils import tf_differentiable
+
+@pytest.fixture
+def start_ray():
+    ray.init()
+
+    yield None
+
+    # teardown
+    ray.shutdown()
+
+@pytest.fixture
+def ray_actor():
+    @ray.remote
+    class Actor(object):
+        """ Ray Actor with TF differentiable functions. """
+
+        def __init__(self):
+            # enable eager execution
+            tf.enable_eager_execution()
+
+        # single input, single output
+        @tf_differentiable(num_return_vals=1)
+        def square(self, x):
+            return x**2
+
+        @tf_differentiable(num_return_vals=1)
+        def cube(self, x):
+            return x**3
+
+        @tf_differentiable(num_return_vals=1)
+        def double(self, x):
+            return 2 * x
+
+
+        # multiple inputs, single output
+        @tf_differentiable(num_return_vals=1)
+        def sum(self, *inputs):
+            return reduce(lambda x, y: x + y, inputs)
+
+        @tf_differentiable(num_return_vals=1)
+        def prod(self, *inputs):
+            return reduce(lambda x, y: x * y, inputs)
+
+        @tf_differentiable(num_return_vals=1)
+        def sum_square_cube(self, x, y):
+            return x**2 + y**3
+
+
+        # single input, multiple outputs
+        @tf_differentiable(num_return_vals=2)
+        def single_in_square_cube(self, x):
+            return x**2, x**3
+
+
+        # multiple inputs, multiple outputs
+        @tf_differentiable(num_return_vals=2)
+        def two_in_square_cube_v1(self, x, y):
+            return x**2, y**3
+
+        @tf_differentiable(num_return_vals=2)
+        def two_in_square_cube_v2(self, x, y):
+            return x**2, x**3
+
+        @tf_differentiable(num_return_vals=5)
+        def f(self, x, y, z):
+            return x**2, y**3, x**4, 2 * z, x * y
+
+        @tf_differentiable(num_return_vals=2)
+        def g(self, x, y, z):
+            return x + y + z, x * y * z
+
+
+        # incorrect number of return values
+        @tf_differentiable(num_return_vals=2)
+        def incorrect_num_return_vals(self, x, y, z):
+            return x**2, y**3, x**4, 2 * z, x * y
+
+    return Actor.remote()
+
+@pytest.fixture
+def dummy_actor():
+    class Actor(object):
+        """ Regular Python version of above Ray Actor. """
+
+        def __init__(self):
+            # enable eager execution
+            tf.enable_eager_execution()
+
+        # single input, single output
+        def square(self, x):
+            return x**2
+
+        def cube(self, x):
+            return x**3
+
+        def double(self, x):
+            return 2 * x
+
+
+        # multiple inputs, single output
+        def sum(self, *inputs):
+            return reduce(lambda x, y: x + y, inputs)
```

Comment:
why not just `sum(inputs)`?

Reply:
generally I think reduce is avoided :) https://stackoverflow.com/questions/13638898/how-to-use-filter-map-and-reduce-in-python-3